File: python/instance.py
# -*- coding: utf-8 -*-
"""
Created on Sun Jan 20 22:42:05 2019

@author: Victor
"""

import io

class Instance:
	def nodeWeight(self,i):
		return self.p[i]

	def edgeDist(self,i,j):
		return self.d[i][j]    

	def hasEdge(self,i,j):
		return self.adj[i][j]

	def __init__(self, filename=None):
		if type(filename) == str:
			with io.open(filename,'r') as f:
				lines = f.readlines()
				
				l = lines[0].replace(" ","")
				self.n = int(l.split("=")[1])
				l = lines[1].replace(" ","")
				self.s = int(l.split("=")[1]) - 1
				l = lines[2].replace(" ","")
				self.t = int(l.split("=")[1]) - 1
				l = lines[3].replace(" ","")
				self.S = int(l.split("=")[1])
				l = lines[4].replace(" ","")
				self.d1 = int(l.split("=")[1])
				l = lines[5].replace(" ","")
				self.d2 = int(l.split("=")[1])
				
				l = lines[6].replace(" ","")
				self.p = []
				for i in l[3:-2].split(","):
					self.p += [int(i)]
				
				l = lines[7].replace(" ","")
				self.ph = []
				for i in l[4:-2].split(","):
					self.ph += [int(i)]
					
				self.initialNodes = range(self.n)
				
				self.m = 0
				
				self.adj = [[0 for i in range(self.n)] for j in range(self.n)]
				self.d = [[0 for i in range(self.n)] for j in range(self.n)]
				self.D = [[0. for i in range(self.n)] for j in range(self.n)]
				self.neighbors = [[] for i in range(self.n)]
				self.predecessors = [[] for i in range(self.n)]
					
				for line in lines[9:]:
					L = line[:-2].split(" ")
					u = int(L[0])-1
					v = int(L[1])-1
					self.adj[u][v] = True
					self.m += 1
					self.d[u][v] = int(L[2])
					self.D[u][v] = float(L[3])
					self.neighbors[u] += [v]
					self.predecessors[v] += [u]
				
	def restrict(self, removedNodes, removedEdges):
		instance = Instance()
		instance.n = 0
		instance.m = 0
		instance.S = self.S
		instance.d1 = self.d1
		instance.d2 = self.d2
		instance.p = []
		instance.ph = []
		instance.initialNodes = []

		instanceToSelf = []        
		selfToInstance = [None for i in range(self.n)]        

		for i in range(self.n):
			if not i in removedNodes:
				selfToInstance[i] = instance.n
				instanceToSelf += [i]
				
				instance.n += 1
				instance.p += [self.p[i]]
				instance.ph += [self.ph[i]]
				instance.initialNodes += [self.initialNodes[i]]
		
		instance.s = selfToInstance[self.s]
		instance.t = selfToInstance[self.t]        
		
		
		instance.adj = [[0 for i in range(instance.n)] for j in range(instance.n)]
		instance.d = [[0 for i in range(instance.n)] for j in range(instance.n)]
		instance.D = [[0. for i in range(instance.n)] for j in range(instance.n)]
		instance.neighbors = [[] for i in range(instance.n)]
		instance.predecessors = [[] for i in range(instance.n)]
		
		for i in range(instance.n):
			for v in self.neighbors[instanceToSelf[i]]:
				if not v in removedNodes and not (instanceToSelf[i],v) in removedEdges:
					instance.neighbors[i] += [selfToInstance[v]]
					instance.predecessors[selfToInstance[v]] += [i]
					instance.adj[i][selfToInstance[v]] = True
					instance.m += 1
					instance.d[i][selfToInstance[v]] = self.d[instanceToSelf[i]][v]
					instance.D[i][selfToInstance[v]] = self.D[instanceToSelf[i]][v]
		
		return instance
		
	def write(self, filename):
		with io.open(filename,'w') as f:
			f.write(u"n = " + str(self.n))
			f.write(u"\n")
			f.write(u"s = " + str(self.s+1))
			f.write(u"\n")
			f.write(u"t = " + str(self.t+1))
			f.write(u"\n")
			f.write(u"S = " + str(self.S))
			f.write(u"\n")
			f.write(u"d1 = " + str(self.d1))
			f.write(u"\n")
			f.write(u"d2 = " + str(self.d2))
			f.write(u"\n")
			f.write(u"p = [" + str(self.p[0]))
			for p in self.p[1:]:
				f.write(u", " + str(p))
			f.write(u"]\n")
			f.write(u"ph = [" + str(self.ph[0]))
			for p in self.ph[1:]:
				f.write(u", " + str(p))
			f.write(u"]\n")
			
			f.write(u"Mat = [\n")
			
			imax = self.n-1
			while (len(self.neighbors[imax]) == 0):
				imax -= 1
			
			for i in range(imax):
				for j in self.neighbors[i]:
					f.write(str(i+1) + u" " + str(j+1) + u" " + str(self.d[i][j]) + u" " + str(self.D[i][j]) + u";\n")
			
			i = imax
			for j in self.neighbors[i][:-1]:
				f.write(str(i+1) + u" " + str(j+1) + u" " + str(self.d[i][j]) + u" " + str(self.D[i][j]) + u";\n")
			
			j = self.neighbors[i][-1]
			f.write(str(i+1) + u" " + str(j+1) + u" " + str(self.d[i][j]) + u" " + str(self.D[i][j]) + u"]\n")
			

File: C++/instance.cpp
#include "instance.h"
#include "node.h"
#include <cstdlib>
#include <cmath>
#include <iostream>
#include <fstream>

using namespace std;

Instance::Instance(char filename[], int verbosity) {
  char readChar;
  int readInt;
  int v1;
  int v2;
  int dist;
  float dev;

  ifstream file(filename);
  if(file) {
    file >> readChar >> readChar >> n;
    if (verbosity >= 1) cout << "n = " << n << endl;

    file >> readChar >> readChar >> s;
    s = s - 1;
    if (verbosity >= 1) cout << "s = " << s << endl;

    file >> readChar >> readChar >> t;
    t = t - 1;
    if (verbosity  >= 1) cout << "t = " << t << endl;

    file >> readChar >> readChar >> S;
    if (verbosity >= 1) cout << "S = " << S << endl;

    file >> readChar >> readInt >> readChar >> d1;
    if (verbosity >= 1) cout << "d1 = " << d1 << endl;

    file >> readChar >> readInt >> readChar >> d2;
    if (verbosity >= 1) cout << "d2 = " << d2 << endl;

	file >> readChar >> readChar >> readChar;
	for (int i = 0; i < n; ++i) {
		file >> readInt >> readChar;
		p.push_back(readInt);
	}
	if (verbosity >= 2) {
		cout << "p = [";
		for (int i = 0; i < n; ++i)
			cout << p[i] << " ";
		cout << "]" << endl;
	}

	file >> readChar >> readChar >> readChar >> readChar;
	for (int i = 0; i < n; ++i) {
		file >> readInt >> readChar;
		ph.push_back(readInt);
	}
	if (verbosity >= 2) {
		cout << "ph = [";
		for (int i = 0; i < n; ++i)
			cout << ph[i] << " ";
		cout << "]" << endl;
	}

    m = 0;
	for (int i = 0; i < n; ++i) {
        initialNodes.push_back(i);
		adj.push_back(vector<bool>(n,false));
		d.push_back(vector<int>(n,0));
		D.push_back(vector<float>(n,0.));
		neighbors.push_back(vector<Node>());
		predecessors.push_back(vector<Node>());
	}

	file >> readChar >> readChar >> readChar >> readChar;

	while (file.peek() != ']') {
		file >> readChar >> v1 >> v2 >> dist >> dev;
		d[v1-1][v2-1] = dist;
		D[v1-1][v2-1] = dev;

		if (dist != 0) {
            adj[v1-1][v2-1] = true;
            m = m + 1;
            neighbors[v1-1].push_back(v2-1);
            predecessors[v2-1].push_back(v1-1);
		}
	}

    file.close();
  }
}

bool Instance::hasEdge(int node1, int node2) const {
	return adj[node1][node2];
}

int Instance::nodeWeight(int node) const {
	return p[node];
}

int Instance::edgeDist(int node1, int node2) const {
	return d[node1][node2];
}

Instance Instance::restrict(std::set<Node> removedNodes, std::set<Edge> removedEdges) {
    Instance instance;
    instance.n = 0;
    instance.S = S;
    instance.d1 = d1;
    instance.d2 = d2;

    std::vector<Node> instanceToSelf;
    std::vector<Node> selfToInstance(n,-1);

    for (unsigned int i = 0; i < n; ++i) {
        if (removedNodes.find(i) == removedNodes.end()) {
            selfToInstance[i] = instance.n;
            instanceToSelf.push_back(i);

            instance.n = instance.n + 1;
            instance.p.push_back(p[i]);
            instance.ph.push_back(ph[i]);
            instance.initialNodes.push_back(initialNodes[i]);
        }
    }

    instance.s = selfToInstance[s];
    instance.t = selfToInstance[t];

    instance.m = 0;
    for (int i = 0; i < instance.n; ++i) {
		instance.adj.push_back(vector<bool>(instance.n,false));
		instance.d.push_back(vector<int>(instance.n,0));
		instance.D.push_back(vector<float>(instance.n,0.));
		instance.neighbors.push_back(vector<Node>());
		instance.predecessors.push_back(vector<Node>());
	}

    for (int i = 0; i < instance.n; ++i) {
        for (unsigned int j = 0; j < neighbors[instanceToSelf[i]].size(); ++j) {
            Node v = neighbors[instanceToSelf[i]][j];
            if (removedNodes.find(v) == removedNodes.end() && removedEdges.find(std::make_pair(instanceToSelf[i],v)) == removedEdges.end()) {
                instance.neighbors[i].push_back(selfToInstance[v]);
                instance.predecessors[selfToInstance[v]].push_back(i);
                instance.adj[i][selfToInstance[v]] = true;
                instance.m = instance.m + 1;
                instance.d[i][selfToInstance[v]] = d[instanceToSelf[i]][v];
                instance.D[i][selfToInstance[v]] = D[instanceToSelf[i]][v];
            }
        }
    }

    return instance;
}
